### PR TITLE
[#11371] Instructor sample data: defer until instructor joins

### LINKED ISSUE #11371: Instructor sample data: defer creation until instructor joins
- **Environment**: V8.0.2, but problem exists since as far back as V6.0.0

**Description of feature/enhancement**

As discussed in https://github.com/TEAMMATES/teammates/issues/11243#issuecomment-898993617.

Only create instructor sample data when the instructor joins.

**Justification**

- If the requesting instructor never uses the system, no sample data is being wasted
- User can supply more information to further customize the sample data, such as timezone (ref: #8787)
- If the account request is persisted in a new entity, this entity can also contain the institute, which means there's no longer a need for institute verification with HMAC value

**Existing similar features and their shortcomings**

Instructor sample data is created when the account is created, which leads to the problems mentioned before.

### PR DESCRIPTION
**Pull request by samuelfangjw**
Fixes #11371 

**Outline of Solution**

- Admin creates a new instructor account by sending a post request to a new `ACCOUNT_REQUEST` endpoint. This invokes the` CreateAccountRequestAction`.
- The action creates a new entity, `AccountRequest`, to store account requests. The account request entity stores the instructors `name`, `email`, `institute` and a generated `registration key`.
- A link is created with the encrypted registration key passed as a url param.
- When an instructor clicks on the link, he is directed to the ~new create account page~ user-join page.
- Once the instructor clicks register, a post request with the registration key is sent to the `ACCOUNT` endpoint.  If the registration key matches the registration key stored in the `AccountsRequestDb`, sample data is created for that instructor.
- The `AccountRequest` is then soft deleted since it is no longer needed.

**UI Changes**
In the user join page, a loading spinner has been added. 

Screenshot 2021-09-09 at 5 36 09 PM

### CODE REVIEW

File: src/e2e/java/teammates/e2e/cases/AdminHomePageE2ETest.java [hunk 1]
```diff
@@ -26,9 +25,6 @@ public void testAll() {
         String name = "AHPUiT Instrúctör WithPlusInEmail";
         String email = "AHPUiT+++_.instr1!@gmail.tmt";
         String institute = "TEAMMATES Test Institute 1";
-        String demoCourseId = "AHPUiT____.instr1_.gma-demo";
-
-        BACKDOOR.deleteCourse(demoCourseId);
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/cases/AdminHomePageE2ETest.java:31]: While you no longer need to delete the sample course anymore, you now need to delete the account request freshly created in this test case.

File: src/main/java/teammates/common/datatransfer/attributes/AccountRequestAttributes.java [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+ * The data transfer object for {@link AccountRequest} entities.
+ */
+public class AccountRequestAttributes extends EntityAttributes<AccountRequest> {
+
+    private String email;
+    private String name;
+    private String institute;
+    private String registrationKey;
+    private transient Instant createdAt;
+    private transient Instant deletedAt;
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/attributes/AccountRequestAttributes.java:22]: Any reason for these fields to be `transient`?

File: src/main/java/teammates/storage/entity/AccountRequest.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+/**
+ * Represents an AccountRequest entity.
+ */
+@Entity
+@Index
+public class AccountRequest extends BaseEntity {
+
+    @Id
+    private String email;
```
Comment (wkurniawan07) [inline src/main/java/teammates/storage/entity/AccountRequest.java:18]: Not wise for this to be primary key. A user can request multiple accounts for different institutes.

File: src/main/java/teammates/ui/output/AccountRequestData.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+/**
+ * Output format of account request data.
+ */
+public class AccountRequestData extends ApiOutput {
+
+    private final String email;
+    private final String name;
+    private final String institute;
+    private final String registrationKey;
+    private final long createdAtTimeStamp;
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/output/AccountRequestData.java:14]: `timestamp` is one word on its own, so there is no need to capitalize the `S`

File: src/main/java/teammates/ui/webapi/CreateAccountRequestAction.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+        JoinLinkData output = new JoinLinkData(joinLink);
+        return new JsonResult(output);
+    }
+
+    /**
+     * Generate unique registration key for the account.
+     * The key contains random elements to avoid being guessed.
+     */
+    private String generateRegistrationKey(String uniqueId) {
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/CreateAccountRequestAction.java:59]: Any reason for creating this here and not in the DB layer, like how the student/instructor keys are created?

File: src/main/java/teammates/ui/webapi/CreateAccountRequestAction.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+                .build();
+
+        try {
+            logic.createOrUpdateAccountRequest(accountRequestAttributes);
+        } catch (InvalidParametersException e) {
+            return new JsonResult(e.getMessage(), HttpStatus.SC_BAD_REQUEST);
+        }
+
+        String joinLink = Config.getFrontEndAppUrl(Const.WebPageURIs.CREATE_ACCOUNT_PAGE)
+                .withRegistrationKey(StringHelper.encrypt(registrationKey))
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/CreateAccountRequestAction.java:44]: On-the-fly encryption is no longer the way to do it as per #11375.

File: src/web/app/app.module.ts [hunk 7]
```diff
@@ -46,6 +46,11 @@ let routes: Routes = [
         component: PublicPageComponent,
         loadChildren: () => import('./user-join-page.module').then((m: any) => m.UserJoinPageModule),
       },
+      {
+        path: 'createaccount',
+        component: PublicPageComponent,
+        loadChildren: () => import('./user-create-account-page.module').then((m: any) => m.UserCreateAccountPageModule),
```
Comment (wkurniawan07) [inline src/web/app/app.module.ts:52]: I'm really not sure why we need an entire new page for this, especially if you look at the duplications. Why not just reuse the user join page? It shouldn't be difficult to differentiate how the page is processed by using some form of query parameter combination.

File: src/test/java/teammates/test/AbstractBackDoor.java [hunk 8]
```diff
@@ -739,4 +739,5 @@ public void deleteCourse(String courseId) {
         }
 
     }
+
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/AbstractBackDoor.java:742]: Unrelated change

File: src/main/java/teammates/storage/api/AccountRequestsDb.java [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+        return load().id(email).now();
+    }
+
+    /**
+     * Deletes an accountRequest.
+     */
+    public void deleteAccountRequest(String email) {
+        assert email != null;
+
+        deleteEntity(Key.create(AccountRequest.class, email));
```
Comment (wkurniawan07) [inline src/main/java/teammates/storage/api/AccountRequestsDb.java:52]: If all deletions are hard deletions, what is the purpose of `deletedAt` field?

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]
This is a very decent initial attempt, but still quite a long way to go. Aware that it's still WIP but I figure many things have to be course-corrected early.

**Comment by damithc** [comment]
Guys, This PR seems to be stalling (no activities for the past 7 days). :snail: :cry:
Hope someone can get it to move forward again soon...

**Comment by samuelfangjw** [comment]
> Guys, This PR seems to be stalling (no activities for the past 7 days). 🐌 😢
> Hope someone can get it to move forward again soon...

So sorry, I'll get to this as soon as possible.

File: src/main/java/teammates/storage/entity/AccountRequest.java (hunk 3 above)
Comment (samuelfangjw) [reply src/main/java/teammates/storage/entity/AccountRequest.java]: Change primary key to take into account both email and institute

File: src/main/java/teammates/ui/webapi/CreateAccountRequestAction.java (hunk 5 above)
Comment (samuelfangjw) [reply src/main/java/teammates/ui/webapi/CreateAccountRequestAction.java]: Moved to db layer

**Comment by samuelfangjw** [comment]
Done with this iteration! Ready for review.

**Changes**
- id field added to account request, which is made from combining instructor email and institute.
- No more on the fly encryption
- reused user join page
- `DeleteAccountRequestAction` and `GetAccountRequestAction` are only used during e2e tests.

**UI Changes**
In the user join page, a loading spinner has been added. 

Screenshot 2021-09-09 at 5 36 09 PM

**Comment by moziliar** [comment]
Reminder of this PR.

**Comment by hhdqirui** [comment]
@wkurniawan07 for your kind review when you have time.

File: src/web/app/user-join-page.component.ts [hunk 10]
```diff
@@ -91,4 +128,28 @@ export class UserJoinPageComponent implements OnInit {
...
+  /**
+   * Creates an account.
+   */
+  createAccount(): void {
+    this.isLoading = true;
+    this.accountService
+      .createAccount(this.key)
+      .pipe(finalize(() => (this.isLoading = false)))
+      .subscribe(
+        (_resp: MessageOutput) => {
```
Comment (wkurniawan07) [inline src/web/app/user-join-page.component.ts:140]: Can just omit the param entirely instead of prefixing with `_`

File: src/web/app/user-join-page.component.ts [hunk 11]
```diff
@@ -43,37 +47,70 @@ export class UserJoinPageComponent implements OnInit {
...
+        this.setupForJoinCourse();
+      }
+    });
+  }
+
+  /**
+   * Setup page for instructor to join course.
+   */
+  setupForJoinCourse(): void {
+    this.courseService.getJoinCourseStatus(this.key, this.entityType).subscribe((resp: JoinStatus) => {
```
Comment (wkurniawan07) [inline src/web/app/user-join-page.component.ts:69]: This part has been refactored in #11421. You probably would find even more common operations/steps with the create account counterpart.

File: src/web/app/user-join-page.component.ts [hunk 12]
```diff
@@ -43,37 +47,70 @@ export class UserJoinPageComponent implements OnInit {
...
+    const nextUrl: string = `${window.location.pathname}${window.location.search.replace(/&/g, '%26')}`;
+    this.authService.getAuthUser(undefined, nextUrl).subscribe((resp: AuthInfo) => {
+      this.userId = resp.user?.id || '';
+
+      if (!resp.user) {
+        window.location.href = `${this.backendUrl}${resp.instructorLoginUrl}`;
+      }
+
+      if (resp.user?.isInstructor) {
+        // User already has instructor account
```
Comment (wkurniawan07) [inline src/web/app/user-join-page.component.ts:109]: This is not sufficient. Create account is an action that transforms an account request to a set of sample data. Just because a user is already an existing instructor in the system, does not mean that the existing instructor corresponds to the requested account creation.

File: src/main/java/teammates/ui/webapi/CreateAccountAction.java [hunk 13]
```diff
@@ -43,17 +61,23 @@ public JsonResult execute() throws InvalidHttpRequestBodyException {
...
+        } catch (EntityDoesNotExistException ednee) {
+            throw new EntityNotFoundException(ednee);
+        } catch (EntityAlreadyExistsException eaee) {
+            throw new InvalidOperationException(eaee);
+        } catch (InvalidParametersException ipe) {
+            throw new InvalidHttpRequestBodyException(ipe);
+        }
+
+        // Delete account request as it is no longer needed
+        logic.deleteAccountRequest(instructorEmail, instructorInstitution);
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/CreateAccountAction.java:78]: Maybe, instead of deleting account request, we should mark it as something like "registered at <time>" instead. This serves two purpose:
- We can keep track/archive the list of all account requests we have, registered or not
- We do wonder at times the time lag between user requesting an account and registering his/her account, so this would provide some insights

Deleting account request still has legitimate usages (e.g. for data used in tests or removing wrongly created entities), but the action here should be different.

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]
Apologies for the delayed review! Solid groundwork done here. Some suggested improvements, and need to sync with the latest `master` branch.

**Comment by samuelfangjw** [comment]
@wkurniawan07. Apologies for the delay, I've made changes according to your comments, please take a look when you have time.

File: src/main/java/teammates/logic/core/AccountRequestsLogic.java [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+     * @throws EntityDoesNotExistException if account request does not exist
+     */
+    public AccountRequestAttributes getAccountRequestForRegistrationKey(String registrationKey)
+            throws EntityDoesNotExistException {
+        AccountRequestAttributes accountRequest = accountRequestsDb
+                .getAccountRequestForRegistrationKey(registrationKey);
+
+        if (accountRequest == null) {
+            throw new EntityDoesNotExistException(
+                    "Account request with registration key " + registrationKey + " does not exist");
```
Comment (wkurniawan07) [inline src/main/java/teammates/logic/core/AccountRequestsLogic.java:71]: Any reason why `EntityDoesNotExistException` is thrown here but not in the previous method where the intent is the exact same (just different in parameters)?

File: src/main/java/teammates/ui/output/AccountRequestData.java [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+ * Output format of account request data.
+ */
+public class AccountRequestData extends ApiOutput {
+
+    private final String email;
+    private final String name;
+    private final String institute;
+    private final String registrationKey;
+    private final Instant registeredAt;
+    private final Instant createdAt;
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/output/AccountRequestData.java:17]: As API output, these should be `Long`. `Instant` is only relevant within the realms of Java.

File: src/test/java/teammates/ui/webapi/CreateAccountActionTest.java [hunk 16]
```diff
@@ -34,34 +29,35 @@ protected String getRequestMethod() {
...
+        verifyHttpParameterFailure();
 
-        ex = verifyHttpRequestBodyFailure(buildCreateRequest(name, null, email));
-        assertEquals("institute cannot be null", ex.getMessage());
+        ______TS("Null parameters");
 
-        ex = verifyHttpRequestBodyFailure(buildCreateRequest(name, institute, null));
-        assertEquals("email cannot be null", ex.getMessage());
+        String[] nullParams = new String[] { Const.ParamsNames.REGKEY, null, };
+        Exception ex = verifyHttpParameterFailure(nullParams);
```
Comment (wkurniawan07) [inline src/test/java/teammates/ui/webapi/CreateAccountActionTest.java:51]: Use the specific exception type here

File: src/test/java/teammates/ui/webapi/CreateAccountRequestActionTest.java [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+    @Test
+    protected void testExecute() throws Exception {
+        loginAsAdmin();
+        String name = "JamesBond";
+        String email = "jamesbond89@gmail.tmt";
+        String institute = "TEAMMATES Test Institute 1";
+
+        ______TS("Null parameters");
+
+        Exception ex = verifyHttpRequestBodyFailure(buildCreateRequest(null, institute, email));
```
Comment (wkurniawan07) [inline src/test/java/teammates/ui/webapi/CreateAccountRequestActionTest.java:38]: Similarly here

File: src/test/java/teammates/ui/webapi/CreateAccountRequestActionTest.java [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+        JsonResult r = getJsonResult(a);
+
+        AccountRequestAttributes accountRequestAttributes = logic.getAccountRequest(email, institute);
+
+        assertEquals(name, accountRequestAttributes.getName());
+        assertEquals(email, accountRequestAttributes.getEmail());
+        assertEquals(institute, accountRequestAttributes.getInstitute());
+        assertNotNull(accountRequestAttributes.getRegistrationKey());
+
+        String joinLink = Config.getFrontEndAppUrl(Const.WebPageURIs.JOIN_PAGE)
```
Comment (wkurniawan07) [inline src/test/java/teammates/ui/webapi/CreateAccountRequestActionTest.java:67]: Can reuse the function defined in `AccountRequestAttributes` here

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]
Good job so far! Still not quite production ready yet, but we're steadily getting there.

**Comment by samuelfangjw** [comment]
So sorry been struggling to find time to work on this recently. I'm more free now, will work on this on monday.

File: src/main/java/teammates/logic/core/AccountRequestsLogic.java (hunk 14 above)
Comment (samuelfangjw) [reply src/main/java/teammates/logic/core/AccountRequestsLogic.java:71]: Will change to throw exception in the previous method as well.

File: src/main/java/teammates/common/datatransfer/attributes/AccountRequestAttributes.java [hunk 19]
```diff
@@ -0,0 +1,281 @@
...
+
+        return errors;
+    }
+
+    @Override
+    public AccountRequest toEntity() {
+        AccountRequest accountRequest = new AccountRequest(getEmail(), getName(), getInstitute());
+
+        if (this.getRegistrationKey() != null) {
+            accountRequest.setRegistrationKey(this.getRegistrationKey());
```
[the remark this answers is not in the document]
Comment (samuelfangjw) [reply src/main/java/teammates/common/datatransfer/attributes/AccountRequestAttributes.java:105]: The default value for registration key and created at is not null. 

In the AccountRequest constructor
```java
  public AccountRequest(String email, String name, String institute) {
      ...
      this.setRegistrationKey(generateRegistrationKey());
      this.setCreatedAt(Instant.now());
  }
```

I believe it is necessary to keep the first two.

File: src/web/app/user-join-page.component.spec.ts [hunk 20]
```diff
@@ -182,3 +186,119 @@ describe('UserJoinPageComponent', () => {
     expect(component.isLoading).toBeFalsy();
   });
 });
+
+describe('UserJoinPageComponent creating account', () => {
```
[the remark this answers is not in the document]
Comment (samuelfangjw) [reply src/web/app/user-join-page.component.spec.ts:190]: Activated route for each describe has different params, one with params for normal request (entitytype, key), another for account request (iscreatingaccount, key). This seemed to be the easiest way to change the route params and seemed like a logical grouping to me. Would this grouping be inappropriate given the above info?

File: src/web/services/account.service.spec.ts [hunk 21]
```diff
@@ -42,13 +42,22 @@ describe('AccountService', () => {
...
   it('should execute POST on account endpoint', () => {
+    const testKey: string = 'testKey';
+    const paramMap: Record<string, string> = {
+      key: testKey,
+    };
+    service.createAccount(testKey);
+    expect(spyHttpRequestService.post).toHaveBeenCalledWith(ResourceEndpoints.ACCOUNT, paramMap);
+  });
+
+  it('should execute POST on account request endpoint', () => {
```
[the remark this answers is not in the document]
Comment (samuelfangjw) [reply src/web/services/account.service.spec.ts:53]: Can I check this point again. The test case is checking if the endpoint is called successfully when the method is called. In my opinion, adding the conditions to the description will not accurately describe the test being carried out here (since the conditions are not checked in the test at all).

File: src/main/java/teammates/ui/output/JoinStatus.java [hunk 22]
```diff
@@ -1,7 +1,7 @@
 package teammates.ui.output;
 
 /**
- * The join status of a course.
+ * The join status of a course or account request.
```
[the remark this answers is not in the document]
Comment (samuelfangjw) [reply src/main/java/teammates/ui/output/JoinStatus.java:4]: > With that in mind, ask yourself this: do you need a dedicated API to get the status of an account request?

To double confirm, are you suggesting we check if instructor is in the course instead?

File: src/main/java/teammates/logic/core/AccountRequestsLogic.java [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+        // No dependency to other logic class
+    }
+
+    /**
+     * Creates or updates an account request.
+     *
+     * @return the created account request
+     * @throws InvalidParametersException if the account request is not valid
+     */
+    public AccountRequestAttributes createOrUpdateAccountRequest(AccountRequestAttributes accountRequestToAdd)
```
[the remark this answers is not in the document]
Comment (samuelfangjw) [reply src/main/java/teammates/logic/core/AccountRequestsLogic.java:35]: Not strictly necessary here, will separate them. Can I check the appropriate behaviour when 2 identical account requests are sent (CreateAccountRequestAction)? Should an error be thrown or the original account request be returned + duplicate email sent.

[the remark this answers is not in the document]
Comment (wkurniawan07) [reply src/main/java/teammates/logic/core/AccountRequestsLogic.java:35]: That is a 409 error.

File: src/main/java/teammates/ui/output/JoinStatus.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (wkurniawan07) [reply src/main/java/teammates/ui/output/JoinStatus.java:4]: > we check if instructor is in the course instead

I don't think I get what you mean by this

[the remark this answers is not in the document]
Comment (samuelfangjw) [reply src/main/java/teammates/ui/output/JoinStatus.java:4]: Sorry could have been clearer. 

You mentioned "An account request is just a special kind of instructor" and "With that in mind, ask yourself this: do you need a dedicated API to get the status of an account request?". Not quite sure what you are trying to hint at here.

**Comment by samuelfangjw** [comment]
@wkurniawan07, so sorry i've updated the PR, left a few comments in the PR review above. Do you mind taking a look when you have some time? No rush!

File: src/main/java/teammates/ui/output/JoinStatus.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (samuelfangjw) [reply src/main/java/teammates/ui/output/JoinStatus.java:4]: @wkurniawan07  Actually would using the `RegkeyValidityData` API be appropriate here?

[the remark this answers is not in the document]
Comment (samuelfangjw) [reply src/main/java/teammates/ui/output/JoinStatus.java:4]: > @wkurniawan07 Actually would using the `RegkeyValidityData` API be appropriate here?

I'll go ahead and implement this for now, do let me know if this is not appropriate.

[the remark this answers is not in the document]
Comment (wkurniawan07) [reply src/main/java/teammates/ui/output/JoinStatus.java:4]: Sorry I just had the chance to review this PR.
What I was trying to hint is reusing `GetCourseJoinStatusAction`.

File: src/web/services/account.service.spec.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (wkurniawan07) [reply src/web/services/account.service.spec.ts:53]: `createAccount` and `createAccountRequest` are very similar, yet very different. Other endpoints have very clear cut difference, thus why just describing "should execute ... on ..." suffices.
Then again it should be fine to keep the description as brief as possible.

File: src/main/java/teammates/ui/output/JoinStatus.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (samuelfangjw) [reply src/main/java/teammates/ui/output/JoinStatus.java:4]: Oh I see. Thanks for the hint.

File: src/main/java/teammates/ui/webapi/GetCourseJoinStatusAction.java [hunk 24]
```diff
@@ -42,12 +46,21 @@ private JsonResult getStudentJoinStatus(String regkey) {
...
+    private JsonResult getInstructorJoinStatus(String regkey, boolean isCreatingAccount) {
+        if (isCreatingAccount) {
+            try {
+                AccountRequestAttributes accountRequest = logic.getAccountRequestForRegistrationKey(regkey);
+                return getJoinStatusResult(accountRequest.getRegisteredAt() != null);
+            } catch (EntityDoesNotExistException ednee) {
+                throw new EntityNotFoundException(ednee);
+            }
+        } else {
+            InstructorAttributes instructor = logic.getInstructorForRegistrationKey(regkey);
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/GetCourseJoinStatusAction.java:58]: If the previous `if` block always terminates, there is no need for `else`. This way we save up one level of indentation. This is known as the "no else return" pattern.

File: src/web/app/user-join-page.component.ts [hunk 25]
```diff
@@ -39,6 +43,12 @@ export class UserJoinPageComponent implements OnInit {
     this.route.queryParams.subscribe((queryParams: any) => {
       this.entityType = queryParams.entitytype;
       this.key = queryParams.key;
+      this.isCreatingAccount = queryParams.iscreatingaccount === 'true';
+
+      // Create cccount request can only come from instructor.
```
Comment (wkurniawan07) [inline src/web/app/user-join-page.component.ts:48]: Typo in "account"

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]
Also, let's add a new E2E test case for when the user registers an account? It should be under `InstructorCourseJoinConfirmationPageE2ETest`.

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]

File: src/main/java/teammates/ui/webapi/CreateAccountAction.java [hunk 26]
```diff
@@ -43,17 +65,24 @@ public JsonResult execute() throws InvalidHttpRequestBodyException {
...
-        EmailWrapper email = emailGenerator.generateNewInstructorAccountJoinEmail(
-                instructorList.get(0).getEmail(), instructorName, joinLink);
-        emailSender.sendEmail(email);
-
-        JoinLinkData output = new JoinLinkData(joinLink);
-        return new JsonResult(output);
+
+        try {
+            logic.joinCourseForInstructor(instructorList.get(0).getKey(), userInfo.id);
+            logic.updateAccountRequest(AccountRequestAttributes
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/CreateAccountAction.java:73]: These two operations should be in separate `try` blocks. We cannot apply a blanket catch-then-rethrow for all the exceptions that are produced here (in general, anywhere within the API layer). For example, neither of the exceptions thrown by the `updateAccountRequest` should be valid here as the entity is guaranteed to exist and valid (the only thing changed is the registration time, which is always valid), thus any of the two being thrown means something is seriously wrong.

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]
As far as the basic functions are concerned, we have all that is needed. Just a few change requested.

However, now that we have introduced a new constraint that there can only be at most one request with unique email + institute combination, we need to also think about how to handle common unhappy paths. For example, there needs to be a way to resend links for account registration, as the previous way of simply recreating the account will no longer work. But this needs not be done on this PR.

File: src/main/java/teammates/ui/webapi/CreateAccountAction.java (hunk 26 above)
Comment (samuelfangjw) [reply src/main/java/teammates/ui/webapi/CreateAccountAction.java:73]: Thanks for the review! Can I ask since "any of the two being thrown means something is seriously wrong", would it be appropriate to log the error using `log.severe()` here?

Comment (wkurniawan07) [reply src/main/java/teammates/ui/webapi/CreateAccountAction.java:73]: You can take a look here: https://github.com/TEAMMATES/teammates/blob/4054fef43227e005ae673a3e0e43b4aae2fd358a/src/main/java/teammates/ui/webapi/JoinCourseAction.java#L66-L84

File: src/main/java/teammates/ui/webapi/CreateAccountAction.java [hunk 27]
```diff
@@ -1,59 +1,99 @@
...
-                instructorList.get(0).getEmail(), instructorName, joinLink);
-        emailSender.sendEmail(email);
-
-        JoinLinkData output = new JoinLinkData(joinLink);
-        return new JsonResult(output);
+
+        try {
+            logic.joinCourseForInstructor(instructorList.get(0).getKey(), userInfo.id);
+        } catch (EntityDoesNotExistException ednee) {
+            throw new EntityNotFoundException(ednee);
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/CreateAccountAction.java:74]: You may want to check whether all of the exceptions thrown here are possible under normal conditions as well. At least `EntityDoesNotExistException` is not for reasons quite obvious.

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]

File: src/main/java/teammates/ui/webapi/CreateAccountAction.java (hunk 27 above)
Comment (samuelfangjw) [reply src/main/java/teammates/ui/webapi/CreateAccountAction.java:74]: I don't believe any of them are possible, will make changes.

File: src/main/java/teammates/ui/webapi/CreateAccountAction.java [hunk 28]
```diff
@@ -1,59 +1,107 @@
...
+                    .withRegisteredAt(Instant.now())
+                    .build());
+        } catch (EntityDoesNotExistException ednee) {
+            // Existence of account request validated before, this exception should not be thrown
+            log.severe("Unexpected error", ednee);
+            return new JsonResult(ednee.getMessage(), HttpStatus.SC_INTERNAL_SERVER_ERROR);
+        } catch (InvalidParametersException ipe) {
+            // There should not be any invalid parameter here
+            log.severe("Unexpected error", ipe);
+            return new JsonResult(ipe.getMessage(), HttpStatus.SC_INTERNAL_SERVER_ERROR);
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/CreateAccountAction.java:101]: The treatment after catch is identical. The catch should therefore be collapsed together i.e. `catch (EntityDoesNotExistException | InvalidParametersException e)`. The explanation comments can always be combined in one paragraph, like:
```java
// EntityDoesNotExistException should not be thrown as existence of account request has been validated before.
// InvalidParametersException should not be thrown as there should not be any invalid parameters.
```

Applies to the previous block as well.

Comment (samuelfangjw) [reply src/main/java/teammates/ui/webapi/CreateAccountAction.java:101]: ok, will update now.

**Review by wkurniawan07 — APPROVED** [review]
LGTM, thanks for the perseverance!

With that being said, we still have quite a bit of work to do.

**Comment by samuelfangjw** [comment]
> LGTM, thanks for the perseverance!
> 
> With that being said, we still have quite a bit of work to do.

Finally! Thanks for the reviews, learnt a lot doing this PR :)
